# Práctica 7
## Herramientas Computacionales
### Autores
* Soto Hernández Derek Anibal        

* Islas Villamar Fernando Jesus    

## Parte-Python

El medicamento administrado a un paciente produce una concentración en la corriente sanguínea dada por: **$$c(t) = Ate^{−t/3}$$** miligramos por mililitros, t horas después de inyectarse A unidades, la máxima concentración segura es de 1mg/ml.

### Actividad 1

* ***¿Qué dosis deberá inyectársele al paciente para alcanzar la máxima concentración segura y cuándo se alcanza esta concentración?***

Hints: 

* Deriva la función e iguale a cero para encontrar un máximo, y use el método de Newton para encontrar la raíz de la función que le queda al igualar a cero. 

* Utilce anterior para determinar la constante $A$.

Tenemos que la concentración del medicamento esta dada por la función:

**$$c(t) = Ate^{−t/3}$$**

Al derivarla obtenemos la función: 

**$$c'(t) = u \cdot v$$**

**$$u = At$$**      

**$$v = e^{-t/3}$$**
  
**$$c'(t) = {-1/3}Ate^{-t/3}+Ae^{-t/3}$$**

Podemos observar que, el punto en t donde la función se hace 0 es t=3

**$$c'(t) = {-1/3}Ate^{-t/3}+Ae^{-t/3} = 0 $$**

Encontramos su factor común:
**$$Ae^{-t/3} \cdot ({-1/3}t + 1) = 0$$**

**$${-1/3}t + 1 = 0$$**

**$$t = -1(-3) = 3$$** 

En esta situación tomamos A como una constante "n".

Asi tenemos que la f de t del método de newton para obtener el tiempo en el que se alcanza la máxima concentración es la derivada de la función original ya que, igualada a 0, y, utilizado su criterio, encontramos el máximo de t(3) de forma analítica:

**$$f(t) = c'(t) = {-1/3}Ate^{-t/3}+Ae^{-t/3}$$**

**$$f'(t) = c''(t)= {1/9}Ate^{-t/3}{-2/3}Ae^{-t/3}$$**


In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [46]:
def Newton_Met(f, t, dfdt, epsilon= 2.0E-8, N= 100):
    
    n=0
    
    while abs(f(t))> epsilon and n<=N:
        t= t-f(t)/dfdt(t)
        n += 1.0
        
    return t, f(t), n

In [47]:
f = lambda t : ((-1/3)*(3*t)*np.e**(-t/3))+(3*np.e**(-t/3))
df= lambda t : ((1/9)*3*t*np.e**(-t/3))+((-2/3)*3*np.e**(-t/3))

In [48]:
seed= 2

out= Newton_Met(f, seed, df)

In [49]:
out

(2.999999994999021, 1.8397574574890996e-09, 4.0)

Para obtener el valor de la dosis en dicho tiempo y la concetración máxima se realizan los pasos

1.- Igualamos la ecuación original a 1 ya que es la que da dicho resultado.
**$$c(t) = Ate^{-t/3} = 1$$**

2.- Despejamos considerando el valor de t = 3.
**$$c(t) = 3Ae^{-t/3}-1 = 0$$**

3.- Como ahora la única incognit es A la ecuación será en función de ella.
**$$c(A) = 3Ae^{-1}-1$$**

4.- Realizamos la derivada.
**$$c'(A) = 3/e$$** 

In [58]:
def Newton_Met(f, A, dfdA, epsilon= 2.0E-8, N= 100):
    
    n=0
    
    while abs(f(A))> epsilon and n<=N:
        A= A-f(A)/dfdA(A)
        n += 1
        
    return A, f(A), n

In [59]:
f = lambda A : (((3*A)*np.e**(-1))-1)
df= lambda A : 3/np.e

In [8]:
seed= 2

out= Newton_Met(f, seed, df)

In [9]:
out

(0.9060939428196817, 0.0, 1)

Con respecto a lo anterior tenemos que la dosis necesaria para llegar a una concentración máxima segura de 1 mg/ml, en 3 horas es de 0.9060... mg. al cual es también el valor de como: **$$A = e/3$$**

### Actividad 2

* ***Una cantidad adicional de medicamento deberá administrarse al paciente después de que la concentración disminuya a 0.25 mg/ml. Determine, con una aproximación al minuto más cercano, cuando debe aplicarse la segunda inyección.***

Para determinar el tiempo al cual debe realizarse la segunda inyección usamos la funición de concentración igualada a 0.25:

**$$0.25 = (e/3)(te^{−t/3})$$**

Una vez que lo tenemos de esta forma, podemos desarrollar una función a partir de la cual obtenemos el tiempo:

**$$f(t)=(e/3)(te^{−t/3})-0.25$$**

Ahora, debemos obtener la raíz que nos indicara el valor en t, al cual se debe realizar la segunda inyección, por lo que derivamos la función:

**$$  $$**


In [10]:
f= lambda t: (((np.e/3)*(t*(np.e**(-t/3))))-(1/4))
df= lambda t: ((-1/3)*(3*t)*np.e**(-t/3))+(3*np.e**(-t/3))

In [11]:
seed= 11

out= Newton_Met(f, seed, df)

In [12]:
out

(11.077903316966399, 1.6388723489413337e-08, 35)

Con lo anterior podemos afirmar que despues de inyectar la primera dosis, se debe aplicar una segunda inyección aproximadamente a los 664.2 minutos (11.07 horas). 

### Actividad 3

* ***Suponiendo que la concentración producida por inyecciones consecutivas es aditiva y que el 75 % de la dosis inyectada originalmente se administra en la segunda inyección. ¿Cuándo será momento de aplicar la tercera inyección?***

Considerando los posibles casos generados, exiten 2 posibles soluciones a esta incógnita de modo que si se sigue el teorema del ejercicio 2 o resolviendo la incógnita con la finalidad de conseguir el máximo de concentración seguro en cada inyección. Desarrollamos las siguientes opciones:

### Opción_1

Al buscar la relación donde la inyección se administra cuando su previa inyección llega al 25% de la concentración (siguiendo dicha relación en el teorema del ejercicio 2 y 3).

Dada: **$$c(t) = 3/4At_1e^{-t_1/3} + At_2e^{-t_2/3}$$**

Donde: **$$3/4At_1e^{-t_1/3}$$** es la ecuación de la segunda dosis y **$$At_2e^{-t_2/3}$$** la ecuación de la primera dosis.
Donde **$$t_1 = 0$$** y **$$t_2 = 11.07$$** y, aplicado en la relación de la ecuación: **$$t_1 = t$$** y **$$t_2 = t + 11.07$$**

Sustituimos los valores en la ecuación y obtenemos nuestra f(x) del método de newton igualado a .25 y luego despejamos:

**$$c(t) = 3/4Ate^{-t/3} + A(t + 11.07)e^{-(t + 11.07)/3} = 0.25$$**
sustituyendo:

**$$f(x) = c(t) = 3/4Ate^{-t/3} + A(t + 11.07)e^{-(t + 11.07)/3} - 0.25 = 0$$**
y derivando:

**$$f'(x) = c'(t) = 3/4(-1/3Ate^{-t/3} + Ae^{-t/3}) + (-1/3Ate^{-(t + 11.07)/3} + Ae^{-(t + 11.07)/3}) + (-1/3(11.07Ae^{-(t + 11.07)/3}))$$**
y aplicar el valor de **$$A = e/3$$**

In [60]:
f= lambda t: ((3/4)*((np.e/3)*t)*np.e**(-t/3)) + ((np.e/3)*(t + 11.07)*np.e**(-(t + 11.07)/3)) -(1/4)
df= lambda t: (3/4)*(((-1/3)*(((np.e/3)*t*np.e**(-t/3)))) + ((np.e/3)*np.e**(-t/3))) + (((-1/3)*((np.e/3)*t*np.e**(-(t + 11.07)/3))) + ((np.e/3)*np.e**(-(t + 11.07)/3))) + ((-1/3)*((11.07)*((np.e/3)*np.e**(-(t + 11.07)/3)))) 

In [70]:
seed= 10


out= Newton_Met(f, seed, df)

In [19]:
out

(10.156022413707621, 1.942890293094024e-15, 3)

De esta manera obtenemos el resultado de t1 el cual es que transcurre de la 2da dosis para la 3ra y, si le agregamos el tiempo que transcurrió de la primera dosis a la segunda, que es 11.07 horas se obtiene el tiempo de 21.22 horas (1273.56 minutos).

### Opción_2

En el caso de idear un teorema donde las concentraciones no cumplan una regla de que indique su administración de dosis luego de que su previa inyección disminuya al 25%, sino al contrario, que esta se inyecte en el momento adecuado para que esta alcance el máximo de concentración en cada inyección, se debe de cumplir cierta relación que cambia por completo todo lo anterior realizado en los ejercicios previos. Siguiendo esta idea, el tiempo de administración de la segunda inyección cambia, por lo cual, debemos de empezar calculando un nuevo tiempo. Para ello seguiremos la misma relación que en la del caso_1, solo que al no saber en que tiempo se administra la segunda inyección lo consideraremos como una constante denominada "***b***". Donde:
**$$t_1 = t0$$** 
y 
**$$t_2 = t + b$$**

Cosiderando que la segunda inyección posee el 75% de la primera inyección se obtiene: 

**$$3/4Ate^{-t/3}$$**

y a la primera dosis inyectada se le debe de considerar su tiempo t y el tiempo en que la segunda inyección se aplicará: 

**$$A(t + b)e^{-(t + b)/3}$$**

Si ahora se genera una función para dicho tiempo:

**$$C(t) = 3/4Ate^{-t/3} + A(t + b)e^{-(t + b)/3}$$**

A dicha función le aplicamos sus condiciones de A como e/3 y a t = 3, con esto igualamos a 1 (la ecuación llega a u máximo de concentración, entonces el tiempo agregado "***b***" será nuestra nuev incógnita para obtener el punto máximo), y sabiendo que se utilizará el método de Newton simplemente despejamos, quedando:

**$$f(x) = C(t) = 3/4Ate^{-t/3} + A(t + b)e^{-(t + b)/3} -1 = 0$$**
a
**$$f(x) = C(b) = 3/4(e/3)3e^{-1} + (e/3)(3 + b)e^{-(3 + b)/3} -1 = 0$$**

Derivamos, y obtenemos: 

**$$f'(x) = C'(b) = 0 + (-1/3)(3(e/3)e^{-(3 + b)/3} + (-1/3)((e/3)be^{-(3 + b)/3} + (e/3)e^{-(3 + b)/3}$$**

In [72]:
def Newton_Met(f, b, dfdb, epsilon= 2.0E-8, N= 100):
    
    n=0
    
    while abs(f(b))> epsilon and n<=N:
        b= b-f(b)/dfdb(b)
        n += 1
        
    return b, f(b), n

In [73]:
f = lambda b: ((3/4)*((np.e/3)*3*np.e**(-1))) + ((np.e/3)*(3 + b)*np.e**(-(3 + b)/3)) -1
df= lambda b: (-(1/3)*(3*(np.e/3)*np.e**(-(3 + b)/3))) + ((-(1/3)*((np.e/3)*b*np.e**(-(3 + b)/3))) + ((np.e/3)*np.e**(-(3 + b)/3)))

In [79]:
seed= 4


out= Newton_Met(f, seed, df)

In [67]:
out

(8.077903524163506, 3.798207082184035e-09, 4)

Este es, el tiempo que debe de reposar el medicamento inyectado hasta inyectar la segunda dosis para que esta última también alcance el máximo de concentración.

Ahora, se calculará el tiempo que transcurre dessde la segunda dosis hasta el momento de inyectar la tercera dosis.

Si consideramos que la primera dosis sigue nuestra función original y la segunda es un 75% (3/4) de la primera y la 3ra debe ser aditiva, con lo cual será el 75% (3/4) de la segunda. Y se sigue la relación entre t1, t2 y t3 donde:

**$$t_1 = t$$**
**$$t_2 = t + c$$** 
**$$t_3 = t + c + 8.07$$**
Tomando como c la incógnita del tiempo que transcurre en la segunda dosis y sumando el tiempo 8.07 que es el de la primera dosis.
Si generamos nuestra función se obtiene:

**$$C(t) = 9/16At_1e^{-t_1/3} + 3/4At_2e^{-t_2/3} + At_3e^{-t_3/3}$$**

Sustituyendo las relaciones de t:
 
**$$C(t) = 9/16Ate^{-t/3} + 3/4A(t + c)e^{-(t + c)/3} + A(t + c + 8.07)e^{-(t + c + 8.07)/3}$$**

Si ahora aplicamos las condiciones de A como e/3, t = 3 e igualamos a 1 sabiendo que es la máxima conectración a obtener, nos da:

**$$ C(3) = (9/16)(e/3)3e^{-1} + (3/4)(e/3)(3 + c)e^{-(3 + c)/3} + (e/3)(3 + c + 8.07)e^{-(3 + c + 8.07)/3} = 1 $$**

Como el valor de t ya está dado, nuestra nueva incógnita es ""c"" por lo tanto, nuestra ecuación será en función de ""c"" y sustituiremos "-1" igualando a 0, la cuál será nuestra función para plaicar el método de Newton: 

**$$f(x) = C(c) = 9/16(e/3)3e^{-1} + 3/4(e/3)(3 + c)e^{-(3 + c)/3} + (e/3)(3 + c + 8.07)e^{-(3 + c + 8.07)/3} -1 = 0$$** 

Y Derivada:

**$$f'(x) = C'(c) = 0 - (1/3)3/4(e/3)3e^{-(3 + c)/3} - (1/3)3/4(e/3)ce^{-(3 + c)/3} + 3/4(e/3)e^{-(3 + c)/3} -(1/3)(e/3)ce^{-(c + 11.07)/3} + (e/3)e^{-(c + 11.07)/3} -(1/3)11.07(e/3)e^{-(c + 11.07)/3}$$**

Simplificando:

**$$f'(x) = C'(c) = -3/4(e/3)e^{-(3 + c)/3} -1/4(e/3)ce^{-(3 + c)/3} +3/4(e/3)e^{-(3 + c)/3} -1/3(e/3)ce^{-(c + 11.07)/3} + (e/3)e^{-(c + 11.07)/3} -1/3(11.07)(e/3)e^{-(c + 11.07)/3}$$**

In [76]:
def Newton_Met(C, c, dCdc, epsilon= 2.0E-8, N= 100):
    
    n=0
    
    while abs(C(c))> epsilon and n<=N:
        c= c-C(c)/dCdc(c)
        n += 1
        
    return c, C(c), n

In [81]:
C = lambda c: ((9/16)*((np.e/3)*3*np.e**(-1))) + ((3/4)*((np.e/3)*(3 + c)*np.e**(-(3 + c)/3))) + ((np.e/3)*((3 + c + 8.07)*np.e**(-(3 + c + 8.07)/3))) - 1
dC =lambda c: ((-3/4)*((np.e/3)*np.e**(-(3 + c)/3))) -((1/4)*((np.e/3)*c*np.e**(-(3 + c)/3))) + ((3/4)*((np.e/3)*np.e**(-(3 + c)/3))) -((1/3)*((np.e/3)*c*np.e**(-(c + 11.07)/3))) +((np.e/3)*np.e**(-(c + 11.07)/3)) -((1/3)*(11.07*(np.e/3)*np.e**(-(c + 11.07)/3)))

In [86]:
seed= 5


out= Newton_Met(C, seed, dC)

In [87]:
out

(5.092425925165482, 3.45148132296913e-09, 2)

Como se observa, 5.09 horas son las que trancurren entre la administración de la segunda dosis y la tercera, si se le suman las 8.07 horas que tomaron entre la primera y la segunda, nos da un total de 13.16 horas hasta la tercera inyección (789.6 minutos).

## Parte LaTeX  

### Actividad 4 

Use la plantilla que se vio en clase para entregar la introducción de su proyecto final. 

### Actividad 5 

Use la plantilla que se vio en clase para entregar parte de la metodología de su proyecto final.